### DISCUSSION TITLE
Simplify Custom UI - remove Caliburn.Micro

### DISCUSSION THREAD

**Discussion by smaudet** [oleg-shilo/wixsharp#1631] [General]
Question/suggestion,

What is Caliburn Micro really providing? Can we just get rid of it?

Possibly it is just the difficulty of debugging UI at an admin level in a mixed-mode setting, possibly it is that windows forms + wpf is just bugged, but Caliburn seems like it is adding another layer of (unwelcome) complexity to the insanity of a stack that is MSI...

As mentioned in another discussion here previously, there is something wrong with the hosted WPF, where under the right circumstances the wpf window can't initialize correctly. Well, now the features dialog has gone and broken (some update, and I was making updates to my installer).

Maybe there is some better way of debugging what is happening in the WPF/WinForms that stuff like bindings don't work and I can't inspect the UI at all (with snoop/microsoft blend, it all crashes and burns horribly), but it seems like Caliburn is just some ancient library that doesn't provide much except another layer of hell to debug...

**Comment by oleg-shilo** [comment]
Sorry for the delay. was engaged with other projects.

It's a good question that does not have a good answer :)

Caliburn Micro is an MVVM that helps orchestrate the dialog initialization under WPF. It is rather a brilliant lightweight lib that worked really well for me for years (and never failed). So I included it in the dialog to minimize the overhead of the manual binding. It does not force the user to use (deal with) it in the user-defined dialogs and it does not prevent the user from using any other lib of this sort (MVVMLight) or going with the manual binding.  

Though there is a serious "however" in this story. 
I wouldn't make the same Caliburn decision today.

The fact that Caliburn introduced the breaking change for their .NET 4.7 forced me to pack two versions of the DLL, to do the runtime dependency resolution. And since there are some hiccups with WPF hosting, having this extra dependency does not help.

Thus I am converting this to the enhancement request and will address it at the next opportunity. It should not be long.
It's a pity, that I will need to resort to manual binding and also impose it on the VS template users for the templates with the dialog sources. But I agree, it's better not to have this dependency.

**Comment by kplatter** [comment]
I am having an issue that may be associated with the 4.7 breaking change that you referred to...   I am getting could not load file or assembly Caliburn.Micro, I have reinstalled Caliburn from NuGet without success...

![image](https://github.com/user-attachments/assets/1434a806-7dff-43a5-83f8-d2c4d80e35b9)

Details:

See the end of this message for details on invoking 
just-in-time (JIT) debugging instead of this dialog box.

************** Exception Text **************
System.IO.FileNotFoundException: Could not load file or assembly 'Caliburn.Micro.Platform, Version=4.0.0.0, Culture=neutral, PublicKeyToken=8e5891231f2ed21f' or one of its dependencies. The system cannot find the file specified.
File name: 'Caliburn.Micro.Platform, Version=4.0.0.0, Culture=neutral, PublicKeyToken=8e5891231f2ed21f'
   at Setup.SplashDialog.Init()
   at WixSharp.UI.WPF.WpfDialogHost.<>c__DisplayClass1_0.<SetDialogContent>b__0(Object s, EventArgs _e)
   at System.Windows.Forms.Form.OnLoad(EventArgs e)
   at System.Windows.Forms.Form.OnCreateControl()
   at System.Windows.Forms.Control.CreateControl(Boolean fIgnoreVisible)
   at System.Windows.Forms.Control.CreateControl()
   at System.Windows.Forms.Control.WmShowWindow(Message& m)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.ScrollableControl.WndProc(Message& m)
   at System.Windows.Forms.ContainerControl.WndProc(Message& m)
   at System.Windows.Forms.Form.WmShowWindow(Message& m)
   at System.Windows.Forms.Form.WndProc(Message& m)
   at System.Windows.Forms.Control.ControlNativeWindow.OnMessage(Message& m)
   at System.Windows.Forms.Control.ControlNativeWindow.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)

WRN: Assembly binding logging is turned OFF.
To enable assembly bind failure logging, set the registry value [HKLM\Software\Microsoft\Fusion!EnableLog] (DWORD) to 1.
Note: There is some performance penalty associated with assembly bind failure logging.
To turn this feature off, remove the registry value [HKLM\Software\Microsoft\Fusion!EnableLog].



************** Loaded Assemblies **************
mscorlib
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.9290.0 built by: NET481REL1LAST_C
    CodeBase: file:///C:/Windows/Microsoft.NET/Framework/v4.0.30319/mscorlib.dll
----------------------------------------
WixToolset.Dtf.WindowsInstaller
    Assembly Version: 4.0.0.0
    Win32 Version: 4.0.1.0
    CodeBase: file:///C:/Users/Keith/AppData/Local/Temp/MSI18071/WixToolset.Dtf.WindowsInstaller.DLL
----------------------------------------
WixSharp.UI
    Assembly Version: 2.6.0.0
    Win32 Version: 2.6.0.0
    CodeBase: file:///C:/Users/Keith/AppData/Local/Temp/MSI18071/WixSharp.UI.DLL
----------------------------------------
WixSharp
    Assembly Version: 2.6.0.0
    Win32 Version: 2.6.0.0
    CodeBase: file:///C:/Users/Keith/AppData/Local/Temp/MSI18071/WixSharp.DLL
----------------------------------------
System.Windows.Forms
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.9251.0 built by: NET481REL1LAST_C
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Windows.Forms/v4.0_4.0.0.0__b77a5c561934e089/System.Windows.Forms.dll
----------------------------------------
System
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.9282.0 built by: NET481REL1LAST_C
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System/v4.0_4.0.0.0__b77a5c561934e089/System.dll
----------------------------------------
System.Drawing
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.9037.0 built by: NET481REL1
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Drawing/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Drawing.dll
----------------------------------------
System.Core
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.9277.0 built by: NET481REL1LAST_B
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Core/v4.0_4.0.0.0__b77a5c561934e089/System.Core.dll
----------------------------------------
Setup
    Assembly Version: 1.0.0.0
    Win32 Version: 1.0.0.0
    CodeBase: file:///C:/Users/Keith/AppData/Local/Temp/MSI18071/Setup.EXE
----------------------------------------
WixSharp.UI.WPF
    Assembly Version: 2.6.0.0
    Win32 Version: 2.6.0.0
    CodeBase: file:///C:/Users/Keith/AppData/Local/Temp/MSI18071/WixSharp.UI.WPF.DLL
----------------------------------------
PresentationFramework
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.9277.0
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/PresentationFramework/v4.0_4.0.0.0__31bf3856ad364e35/PresentationFramework.dll
----------------------------------------
WindowsBase
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.9277.0 built by: NET481REL1LAST_B
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/WindowsBase/v4.0_4.0.0.0__31bf3856ad364e35/WindowsBase.dll
----------------------------------------
PresentationCore
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.9277.0 built by: NET481REL1LAST_B
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_32/PresentationCore/v4.0_4.0.0.0__31bf3856ad364e35/PresentationCore.dll
----------------------------------------
System.Xaml
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.9277.0 built by: NET481REL1LAST_B
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Xaml/v4.0_4.0.0.0__b77a5c561934e089/System.Xaml.dll
----------------------------------------
System.Xml.Linq
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.9037.0 built by: NET481REL1
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Xml.Linq/v4.0_4.0.0.0__b77a5c561934e089/System.Xml.Linq.dll
----------------------------------------
System.Xml
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.9037.0 built by: NET481REL1
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Xml/v4.0_4.0.0.0__b77a5c561934e089/System.Xml.dll
----------------------------------------
System.Configuration
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.9037.0 built by: NET481REL1
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Configuration/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Configuration.dll
----------------------------------------
Accessibility
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.9037.0 built by: NET481REL1
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/Accessibility/v4.0_4.0.0.0__b03f5f7f11d50a3a/Accessibility.dll
----------------------------------------
Newtonsoft.Json
    Assembly Version: 13.0.0.0
    Win32 Version: 13.0.3.27908
    CodeBase: file:///C:/Users/Keith/AppData/Local/Temp/MSI18071/Newtonsoft.Json.DLL
----------------------------------------
System.Net.Http
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.9037.0 built by: NET481REL1
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Net.Http/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Net.Http.dll
----------------------------------------
PresentationFramework.Aero2
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.9037.0 built by: NET481REL1
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/PresentationFramework.Aero2/v4.0_4.0.0.0__31bf3856ad364e35/PresentationFramework.Aero2.dll
----------------------------------------
WindowsFormsIntegration
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.9195.0 built by: NET481REL1LAST_B
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/WindowsFormsIntegration/v4.0_4.0.0.0__31bf3856ad364e35/WindowsFormsIntegration.dll
----------------------------------------
PresentationFramework-SystemXmlLinq
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.9037.0
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/PresentationFramework-SystemXmlLinq/v4.0_4.0.0.0__b77a5c561934e089/PresentationFramework-SystemXmlLinq.dll
----------------------------------------
PresentationFramework-SystemXml
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.9037.0
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/PresentationFramework-SystemXml/v4.0_4.0.0.0__b77a5c561934e089/PresentationFramework-SystemXml.dll
----------------------------------------

************** JIT Debugging **************
To enable just-in-time (JIT) debugging, the .config file for this
application or computer (machine.config) must have the
jitDebugging value set in the system.windows.forms section.
The application must also be compiled with debugging
enabled.

For example:

<configuration>
    <system.windows.forms jitDebugging="true" />
</configuration>

When JIT debugging is enabled, any unhandled exception
will be sent to the JIT debugger registered on the computer
rather than be handled by this dialog box.

**Comment by oleg-shilo** [comment]
PLeas emove to the latest version of WixSharp and you will not have dependency on Calibutrn.Micro. Just raw WPF.

**Comment by kplatter** [comment]
Ty, I already had latest version of WixSharp but had to replace all of my Dialogs which wasn't obvious at first.  I am still Cleaning that up but think it is working as planned.  At least the "could not load file or assembly Caliburn.Micro" error message is gone :)

**Reply by oleg-shilo** [reply]
Yeah, I expected that. 
Since your dialogs are created from the prev version of templates you did have the Caliburtn refs that need to be removed manually. 
The new versions of VS templates (since Sep 21, 2024) do not all Caliburn refs. So we are good. 

But yes you need to remove the refs manually. 
It's easy though. Just remove the Caliburn nuget pack and the compiler will tell you exactly where you need to change the code.
